### ISSUE TITLE
Generate activity diagrams

### ISSUE THREAD

**Issue by jermon** [jfeliu007/goplantuml#105]
Is should be possible to generate activity diagrams from the code in methods and functions. 
https://plantuml.com/activity-diagram-beta

**Comment by jfeliu007** [comment]
It would be nice. I have to give that some thought.

**Comment by hitzhangjie** [comment]
I am trying on this feature, it's a little complex, not that so complex.

Before showing some samples, I want to share my thinkings.
1. which language construct or activities should we visualize?
    I think OOP is good. 
    Maybe what we cares most is the Participants and Communication.
    Besides, we cares the condition branches. like when something happens? If it happens, what we should do?
2. what communication stands for? 
    - method call (method has receiver type): object sends message to another object
    - package exported function call: this maybe not a message passing, but it still looks important for us.
    - RPC or network activity.
    - etc.

    ps: After I do some trying on this, I think it would be better if we could specify which activity we should focus on.
    
3. how should we visualize these constructs or activites?
    - plantuml: plantuml is an option, but maybe we should look for better solutions.
      plantuml generates an static picture, sometime it is enough, maybe most time, it's not.
      when we reading code or trying to understand code, what we focus changes as we navigate, a little like the IDE supports code folding feature. 
      some syntax like if-else/switch-case/for-loop rendered by plantuml is a little hard to understand, the alt-else/loop is not clear enough.

------

some samples:

1. IfStmt

```go
func TestIfElse() {

	num := 1
	if num == 1 {
		fmt.Println("1")
	} else if num == 2 {
		fmt.Println("2")
	} else {
		fmt.Println("others")
	}
}
```

and relevant puml:

```puml
@startuml
participant ifstmt
alt ../../testdata/ast/ifelse.go:8:5
	ifstmt->ifstmt
	note right: stmt ../../testdata/ast/ifelse.go:9:3
else ../../testdata/ast/ifelse.go:10:9
	alt ../../testdata/ast/ifelse.go:10:12
		ifstmt->ifstmt
		note right: stmt ../../testdata/ast/ifelse.go:11:3
	else ../../testdata/ast/ifelse.go:12:9
		ifstmt->ifstmt
		note right: stmt ../../testdata/ast/ifelse.go:13:3
	end
end
@enduml
```

rendered picture:

![image](https://user-images.githubusercontent.com/3725760/103734261-8c6f0f00-5026-11eb-9941-8a385a39e352.png)

ForStmt

```go
package ast

import "fmt"

func TestForStmt() {
	for i := 0; i < 10; i++ {
		fmt.Println("hello world")
	}
}
```

relevant puml:

```puml
@startuml
participant forStmt
group ForStmt
	forStmt->forStmt
	note right: init ../../testdata/ast/for.go:6:6
	loop ../../testdata/ast/for.go:6:14
		forStmt->forStmt
		note right: body stmt: ../../testdata/ast/for.go:7:3
		forStmt->forStmt
		note right: post ../../testdata/ast/for.go:6:22
	end
end
@enduml
```

rendered picture:

![image](https://user-images.githubusercontent.com/3725760/103734355-c50ee880-5026-11eb-91af-0c220eb67a55.png)

**Comment by jermon** [comment]
Producing interaction diagrams  are one thing. Producing activity diagrams are another.  
The purpose of the interaction diagram is to capture the dynamic aspect of a system. Who is taking to who?
The purpose of activity diagram is basically a flowchart to represent the flow from one activity to another activity. The activity can be described as an operation of the system.
See [UML - Activity Diagrams](https://www.tutorialspoint.com/uml/uml_activity_diagram.htm)

**Comment by hitzhangjie** [comment]
@jermon Yes, sequence diagram and activity diagram are two different diagrams.

I want to share my learned experience here. If you think they're completely irrelevant. Le't me know. I'll delete this to avoid letting others misunderstood.

**Comment by jfeliu007** [comment]
I need to see how to make sense of code into an activity diagram for it to be helpful. I can certainly follow the flow of the codes and calls to make activity diagram elements and interactions. I need to l ook more into this.

**Comment by jfeliu007** [comment]
If you have seen before any articles that point in the right direction and would like to share them here, that can also help. Thanks for the suggestion.
